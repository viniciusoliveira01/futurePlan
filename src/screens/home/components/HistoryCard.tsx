import React from 'react';
import ProgressBarAnimated from 'react-native-progress-bar-animated';
import colors from '@config/colors';
import {
  ContentContainer,
  ContentRow,
  DateContainer,
  DateView,
  DateText,
  HistoryCardContainer,
  GoalText,
} from './HistoryCard.style';
import { HistoryCardProps } from '@src/types/historyCard';

export const calculateProgressPercent = (
  valueNow: number,
  goal: number,
): number => (valueNow / goal) * 100;

const HistoryCard = ({ title, valueNow, goal, dateYear }: HistoryCardProps) => {
  const progressPercent = calculateProgressPercent(valueNow, goal);
  const isProgressCompleted =
    progressPercent === 100 ? colors.green : colors.purple;

  return (
    <HistoryCardContainer>
      <DateContainer>
        <DateView progressPercent={progressPercent}>
          <DateText>{dateYear}</DateText>
        </DateView>
      </DateContainer>
      <ContentContainer>
        <ContentRow>
          <GoalText progressPercent={progressPercent}>{title}</GoalText>
          <GoalText progressPercent={progressPercent}>
            {progressPercent}%
          </GoalText>
        </ContentRow>
        <ProgressBarAnimated
          width={300}
          value={progressPercent}
          backgroundColor={isProgressCompleted}
        />
      </ContentContainer>
    </HistoryCardContainer>
  );
};

export default HistoryCard;
