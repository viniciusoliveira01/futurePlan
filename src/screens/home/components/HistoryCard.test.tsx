import React from 'react';
import { render } from '@testing-library/react-native';
import HistoryCard, { calculateProgressPercent } from './HistoryCard';

describe('History Card', () => {
  it('Should render History Card', () => {
    const { toJSON } = render(
      <HistoryCard
        title="Atingir 20000000"
        valueNow={100000}
        goal={20000000}
        dateYear="2050"
      />,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('Should show title, year and percentage', () => {
    const { getByText } = render(
      <HistoryCard
        title="Atingir 20000000"
        valueNow={100000}
        goal={20000000}
        dateYear="2050"
      />,
    );

    expect(getByText(/Atingir 20000000/i)).toBeTruthy();
    expect(getByText(/2050/i)).toBeTruthy();
    expect(getByText(/0.5%/i)).toBeTruthy();
  });

  it('Should calculate the percentage based on valueNow and Goal', () => {
    const valueNow = 72000;
    const goal = 100000;
    const progressPercent = calculateProgressPercent(valueNow, goal);

    expect(progressPercent).toBe(72);
  });

  it('Should calculate the percentage based on valueNow and Goal passed as strings', () => {
    const valueNow = '9000';
    const goal = '100000';
    const progressPercent = calculateProgressPercent(valueNow, goal);

    expect(progressPercent).toBe(9);
  });
});
