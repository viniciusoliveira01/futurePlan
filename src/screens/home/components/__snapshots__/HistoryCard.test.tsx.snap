// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History Card Should render History Card 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFF",
        "borderRadius": 5,
        "elevation": 1,
        "marginHorizontal": 20,
        "marginTop": 40,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.27,
        "shadowRadius": 4.65,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
      ]
    }
  >
    <View
      progressPercent={0.5}
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#7025FF",
            "borderRadius": 5,
            "justifyContent": "center",
            "marginTop": -10,
            "width": 50,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "#FFF",
              "fontSize": 10,
              "fontWeight": "700",
              "marginBottom": 5,
              "marginHorizontal": 10,
              "marginLeft": 5,
              "marginRight": 5,
              "marginTop": 5,
            },
          ]
        }
      >
        2050
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "paddingBottom": 20,
          "paddingLeft": 20,
          "paddingRight": 20,
          "paddingTop": 20,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
            "width": "100%",
          },
        ]
      }
    >
      <Text
        progressPercent={0.5}
        style={
          Array [
            Object {
              "color": "#8c8c8c",
              "fontSize": 14,
              "fontWeight": "700",
              "marginBottom": 15,
            },
          ]
        }
      >
        Atingir 20000000
      </Text>
      <Text
        progressPercent={0.5}
        style={
          Array [
            Object {
              "color": "#8c8c8c",
              "fontSize": 14,
              "fontWeight": "700",
              "marginBottom": 15,
            },
          ]
        }
      >
        0.5
        %
      </Text>
    </View>
    <View
      style={
        Object {
          "borderColor": "#C8CCCE",
          "borderRadius": 6,
          "borderWidth": 1,
          "height": 15,
          "width": 300,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#7025FF",
            "borderRadius": 6,
            "height": 13,
            "width": 0,
          }
        }
      />
    </View>
  </View>
</View>
`;
